File: app/src/main/java/com/jarvis/mobile/live/GeminiLiveClient.kt
package com.jarvis.mobile.live

import android.util.Base64
import android.util.Log
import com.google.gson.JsonArray
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import okio.ByteString
import java.util.concurrent.TimeUnit

class GeminiLiveClient(
    private val apiKey: String,
    private val model: String  = "models/gemini-2.5-flash-native-audio-preview-12-2025",
    private val voiceName: String = "Puck",
) {
    interface Listener {
        fun onSetupComplete()
        fun onAudioChunk(pcm: ByteArray)
        fun onInputTranscript(text: String)
        fun onOutputTranscript(text: String)
        fun onTurnComplete()
        fun onToolCall(calls: List<ToolCall>)
        fun onError(reason: String)
        fun onClosed()
    }

    data class ToolCall(val id: String, val name: String, val args: Map<String, Any?>)

    private val http = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(0, TimeUnit.MILLISECONDS)
        .pingInterval(20, TimeUnit.SECONDS)
        .build()

    private var ws: WebSocket? = null
    private var listener: Listener? = null

    fun connect(systemInstruction: String, toolsBlock: JsonObject, listener: Listener) {
        this.listener = listener
        val url = "wss://generativelanguage.googleapis.com/ws/" +
                  "google.ai.generativelanguage.v1beta.GenerativeService.BidiGenerateContent?key=$apiKey"
        val req = Request.Builder().url(url).build()
        ws = http.newWebSocket(req, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                Log.i(TAG, "WebSocket open")
                sendSetup(webSocket, systemInstruction, toolsBlock)
            }
            override fun onMessage(webSocket: WebSocket, text: String) = handleJson(text)
            override fun onMessage(webSocket: WebSocket, bytes: ByteString) = handleJson(bytes.utf8())
            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                Log.e(TAG, "WebSocket failure", t)
                listener.onError(t.message ?: "ws failure")
            }
            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                Log.i(TAG, "WebSocket closed: $code $reason")
                listener.onClosed()
            }
        })
    }

    fun close() {
        ws?.close(1000, "client close")
        ws = null
    }

    fun sendMicChunk(pcm: ByteArray) {
        val b64 = Base64.encodeToString(pcm, Base64.NO_WRAP)
        val msg = JsonObject().apply {
            add("realtime_input", JsonObject().apply {
                add("media_chunks", JsonArray().apply {
                    add(JsonObject().apply {
                        addProperty("mime_type", "audio/pcm;rate=16000")
                        addProperty("data", b64)
                    })
                })
            })
        }
        ws?.send(msg.toString())
    }

    fun sendUserText(text: String, turnComplete: Boolean = true) {
        val msg = JsonObject().apply {
            add("client_content", JsonObject().apply {
                add("turns", JsonArray().apply {
                    add(JsonObject().apply {
                        addProperty("role", "user")
                        add("parts", JsonArray().apply {
                            add(JsonObject().apply { addProperty("text", text) })
                        })
                    })
                })
                addProperty("turn_complete", turnComplete)
            })
        }
        ws?.send(msg.toString())
    }

    fun sendToolResponses(responses: List<Pair<ToolCall, String>>) {
        val arr = JsonArray()
        for ((call, result) in responses) {
            arr.add(JsonObject().apply {
                addProperty("id", call.id)
                addProperty("name", call.name)
                add("response", JsonObject().apply {
                    addProperty("result", result)
                })
            })
        }
        val msg = JsonObject().apply {
            add("tool_response", JsonObject().apply {
                add("function_responses", arr)
            })
        }
        ws?.send(msg.toString())
    }

    private fun sendSetup(webSocket: WebSocket, systemInstruction: String, toolsBlock: JsonObject) {
        val setup = JsonObject().apply {
            add("setup", JsonObject().apply {
                addProperty("model", model)
                add("generation_config", JsonObject().apply {
                    add("response_modalities", JsonArray().apply { add("AUDIO") })
                    add("speech_config", JsonObject().apply {
                        add("voice_config", JsonObject().apply {
                            add("prebuilt_voice_config", JsonObject().apply {
                                addProperty("voice_name", voiceName)
                            })
                        })
                    })
                })
                add("system_instruction", JsonObject().apply {
                    addProperty("role", "system")
                    add("parts", JsonArray().apply {
                        add(JsonObject().apply { addProperty("text", systemInstruction) })
                    })
                })
                add("tools", JsonArray().apply { add(toolsBlock) })
                add("output_audio_transcription", JsonObject())
                add("input_audio_transcription", JsonObject())
            })
        }
        webSocket.send(setup.toString())
    }

    private fun handleJson(raw: String) {
        val l = listener ?: return
        val root = runCatching { JsonParser.parseString(raw).asJsonObject }.getOrNull() ?: return

        if (root.has("setupComplete")) {
            l.onSetupComplete()
            return
        }

        root.getAsJsonObject("serverContent")?.let { sc ->
            sc.getAsJsonObject("modelTurn")?.getAsJsonArray("parts")?.forEach { p ->
                val part = p.asJsonObject
                part.getAsJsonObject("inlineData")?.let { inline ->
                    val data = inline.get("data")?.asString ?: return@let
                    val pcm = Base64.decode(data, Base64.DEFAULT)
                    l.onAudioChunk(pcm)
                }
            }
            sc.getAsJsonObject("inputTranscription")?.get("text")?.asString?.let {
                if (it.isNotBlank()) l.onInputTranscript(it)
            }
            sc.getAsJsonObject("outputTranscription")?.get("text")?.asString?.let {
                if (it.isNotBlank()) l.onOutputTranscript(it)
            }
            if (sc.get("turnComplete")?.asBoolean == true) {
                l.onTurnComplete()
            }
        }

        root.getAsJsonObject("toolCall")?.getAsJsonArray("functionCalls")?.let { arr ->
            val calls = mutableListOf<ToolCall>()
            for (e in arr) {
                val o = e.asJsonObject
                val argsObj = o.getAsJsonObject("args") ?: JsonObject()
                val argsMap = mutableMapOf<String, Any?>()
                for ((k, v) in argsObj.entrySet()) {
                    argsMap[k] = when {
                        v.isJsonPrimitive && v.asJsonPrimitive.isString  -> v.asString
                        v.isJsonPrimitive && v.asJsonPrimitive.isBoolean -> v.asBoolean
                        v.isJsonPrimitive && v.asJsonPrimitive.isNumber  -> v.asDouble
                        v.isJsonNull -> null
                        else -> v.toString()
                    }
                }
                calls += ToolCall(
                    id   = o.get("id")?.asString ?: "",
                    name = o.get("name")?.asString ?: "",
                    args = argsMap
                )
            }
            if (calls.isNotEmpty()) l.onToolCall(calls)
        }
    }

    companion object { private const val TAG = "GeminiLive" }
}

File: ios/Sources/Live/GeminiLiveClient.swift
import Foundation

/// WebSocket client for Gemini Live `BidiGenerateContent`.
/// Mirrors live/GeminiLiveClient.kt in the Android module.
final class GeminiLiveClient: NSObject {

    struct ToolCall {
        let id: String
        let name: String
        let args: [String: Any]
    }

    protocol Listener: AnyObject {
        func onSetupComplete()
        func onAudioChunk(_ pcm: Data)
        func onInputTranscript(_ text: String)
        func onOutputTranscript(_ text: String)
        func onTurnComplete()
        func onToolCall(_ calls: [ToolCall])
        func onError(_ reason: String)
        func onClosed()
    }

    private let apiKey: String
    private let model: String
    private let voiceName: String

    private var session: URLSession!
    private var task: URLSessionWebSocketTask?
    private weak var listener: Listener?
    private var pendingSystemInstruction: String?
    private var pendingTools: [String: Any]?
    private var isClosed = false

    init(
        apiKey: String,
        model: String = "models/gemini-2.5-flash-native-audio-preview-12-2025",
        voiceName: String = "Puck"
    ) {
        self.apiKey = apiKey
        self.model = model
        self.voiceName = voiceName
        super.init()
        let cfg = URLSessionConfiguration.default
        cfg.timeoutIntervalForRequest = 0   // long-lived
        cfg.waitsForConnectivity = true
        self.session = URLSession(configuration: cfg, delegate: self, delegateQueue: nil)
    }

    func connect(systemInstruction: String, tools: [String: Any], listener: Listener) {
        self.listener = listener
        self.pendingSystemInstruction = systemInstruction
        self.pendingTools = tools
        self.isClosed = false

        let urlStr = "wss://generativelanguage.googleapis.com/ws/" +
                     "google.ai.generativelanguage.v1beta.GenerativeService.BidiGenerateContent" +
                     "?key=\(apiKey)"
        guard let url = URL(string: urlStr) else {
            listener.onError("invalid ws url")
            return
        }
        let req = URLRequest(url: url)
        let t = session.webSocketTask(with: req)
        self.task = t
        t.resume()
        receiveLoop()
    }

    func close() {
        isClosed = true
        task?.cancel(with: .goingAway, reason: nil)
        task = nil
    }

    // MARK: - Outgoing

    func sendMicChunk(_ pcm: Data) {
        let payload: [String: Any] = [
            "realtime_input": [
                "media_chunks": [
                    [
                        "mime_type": "audio/pcm;rate=16000",
                        "data": pcm.base64EncodedString()
                    ]
                ]
            ]
        ]
        sendJSON(payload)
    }

    func sendUserText(_ text: String, turnComplete: Bool = true) {
        let payload: [String: Any] = [
            "client_content": [
                "turns": [
                    [
                        "role": "user",
                        "parts": [["text": text]]
                    ]
                ],
                "turn_complete": turnComplete
            ]
        ]
        sendJSON(payload)
    }

    func sendToolResponses(_ responses: [(ToolCall, String)]) {
        let arr: [[String: Any]] = responses.map { call, result in
            [
                "id":   call.id,
                "name": call.name,
                "response": ["result": result]
            ]
        }
        let payload: [String: Any] = [
            "tool_response": ["function_responses": arr]
        ]
        sendJSON(payload)
    }

    // MARK: - Incoming

    private func receiveLoop() {
        task?.receive { [weak self] result in
            guard let self else { return }
            switch result {
            case .failure(let err):
                if !self.isClosed {
                    self.listener?.onError(err.localizedDescription)
                }
                self.listener?.onClosed()
            case .success(let msg):
                switch msg {
                case .string(let s):
                    self.handle(json: s)
                case .data(let d):
                    if let s = String(data: d, encoding: .utf8) { self.handle(json: s) }
                @unknown default:
                    break
                }
                self.receiveLoop()
            }
        }
    }

    private func handle(json raw: String) {
        guard let l = listener,
              let data = raw.data(using: .utf8),
              let root = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any]
        else { return }

        if root["setupComplete"] != nil {
            l.onSetupComplete()
            return
        }

        if let sc = root["serverContent"] as? [String: Any] {
            if let mt = sc["modelTurn"] as? [String: Any],
               let parts = mt["parts"] as? [[String: Any]] {
                for p in parts {
                    if let inline = p["inlineData"] as? [String: Any],
                       let b64 = inline["data"] as? String,
                       let pcm = Data(base64Encoded: b64) {
                        l.onAudioChunk(pcm)
                    }
                }
            }
            if let it = sc["inputTranscription"] as? [String: Any],
               let t = it["text"] as? String, !t.trimmingCharacters(in: .whitespaces).isEmpty {
                l.onInputTranscript(t)
            }
            if let ot = sc["outputTranscription"] as? [String: Any],
               let t = ot["text"] as? String, !t.trimmingCharacters(in: .whitespaces).isEmpty {
                l.onOutputTranscript(t)
            }
            if let tc = sc["turnComplete"] as? Bool, tc {
                l.onTurnComplete()
            }
        }

        if let tc = root["toolCall"] as? [String: Any],
           let arr = tc["functionCalls"] as? [[String: Any]] {
            let calls: [ToolCall] = arr.compactMap { o in
                let args = (o["args"] as? [String: Any]) ?? [:]
                return ToolCall(
                    id:   (o["id"]   as? String) ?? "",
                    name: (o["name"] as? String) ?? "",
                    args: args
                )
            }
            if !calls.isEmpty { l.onToolCall(calls) }
        }
    }

    // MARK: - Setup

    fileprivate func sendSetup() {
        guard let sys = pendingSystemInstruction, let tools = pendingTools else { return }
        let setup: [String: Any] = [
            "setup": [
                "model": model,
                "generation_config": [
                    "response_modalities": ["AUDIO"],
                    "speech_config": [
                        "voice_config": [
                            "prebuilt_voice_config": ["voice_name": voiceName]
                        ]
                    ]
                ],
                "system_instruction": [
                    "role": "system",
                    "parts": [["text": sys]]
                ],
                "tools": [tools],
                "output_audio_transcription": [:],
                "input_audio_transcription": [:]
            ]
        ]
        sendJSON(setup)
    }

    private func sendJSON(_ obj: [String: Any]) {
        guard let data = try? JSONSerialization.data(withJSONObject: obj, options: []),
              let s = String(data: data, encoding: .utf8)
        else { return }
        task?.send(.string(s)) { [weak self] err in
            if let err, !(self?.isClosed ?? true) {
                self?.listener?.onError(err.localizedDescription)
            }
        }
    }
}

extension GeminiLiveClient: URLSessionWebSocketDelegate {
    func urlSession(_ session: URLSession, webSocketTask: URLSessionWebSocketTask,
                    didOpenWithProtocol protocol: String?) {
        sendSetup()
    }
    func urlSession(_ session: URLSession, webSocketTask: URLSessionWebSocketTask,
                    didCloseWith closeCode: URLSessionWebSocketTask.CloseCode,
                    reason: Data?) {
        listener?.onClosed()
    }
}
